syntax = "proto2";

package chews;

message NodeConfig {
  // Virtual Machine Instance size, e.g. 'n1-standard-4'.  Safe to
  // change after the workstation has been created.
  optional string size = 1;
}

message VolumeConfig {
  // Volume name.  Only needs to unique within CwsWorkstation.
  // Examples: 'boot', 'data'.
  optional string name = 1;

  // Volume size in GB.
  optional int32 size = 2;

  // Volume type.  Example: 'pd-standard' (HDD-backed), 'pd-ssd'
  // (SSD-backed).  Safe to change after the workstation as been
  // created.
  optional string volume_type = 3;

  // Number of snapshots to keep.  Must be at least 1.  Safe to change
  // after the workstation has been created.
  //
  // Note that although GCP implements differential snapshots, the
  // snapshots created by Chews cannot shared blocks since the
  // original disks are deleted between creation of snapshots.  This
  // means that whenever a new snapshot is created, it cannot
  // correlate existing snapshots as previous versions of the disk.
  // As a result, each snapshot is full sized.
  optional int32 max_snapshots = 4;
}

message CwsConfig {
  // Workstation name.  Must be unique within the GCP project.
  optional string name = 1;

  // Workstation configuration.
  optional NodeConfig node = 2;

  // Volumes to be attached to this workstation.  There must be at
  // least one volume.  The first volume is used as the boot disk.
  repeated VolumeConfig volumes = 3;

  // GCP location, e.g. 'us-west1-a'.
  optional string location = 4;

  // Image family used on creation.  Examples: 'debian-9',
  // 'windows-2016'.  See
  // https://console.cloud.google.com/compute/images
  optional string image_family = 5;
}

message Config {
  // Must be set to 'GCE' for now.
  optional string provider = 1;

  // GCP project name.
  optional string project = 2;

  // List of workstations.
  repeated CwsConfig cloud_workstations = 3;
}